samza-core/src/main/java/org/apache/samza/clustermanager/ContainerProcessManager.java

```diff
@@ -513,12 +509,23 @@ void onResourceCompletedWithUnknownStatus(SamzaResourceStatus resourceStatus, St
     int retryCount = clusterManagerConfig.getContainerRetryCount();
     int retryWindowMs = clusterManagerConfig.getContainerRetryWindowMs();
     int currentFailCount;
+    boolean retryContainerRequest = true;
 
     if (retryCount == 0) {
       LOG.error("Processor ID: {} (current Container ID: {}) failed, and retry count is set to 0, " +
           "so shutting down the application master and marking the job as failed.", processorId, containerId);
 
-      jobFailureCriteriaMet = true;
+      // Failure criteria met only if failed containers can fail the job.
+      jobFailureCriteriaMet = clusterManagerConfig.getContainerFailJobAfterRetries();
+      if (jobFailureCriteriaMet) {
+        LOG.error("Processor ID: {} (current Container ID: {}) failed, and retry count is set to 0, " +
+            "so shutting down the application master and marking the job as failed.", processorId, containerId);
+      } else {
+        LOG.error("Processor ID: {} (current Container ID: {}) failed, and retry count is set to 0, " +
+            "marking the container as completed and will not retry request.", processorId, containerId);
+        incrementCompletedProcessorsAndUpdateState(state);
```

Comment:
Perhaps we dont need the call here, since 
a. this will increment the "succeeded" container count regardless of the container's exit status. 
b. it increments the "suceeded" container count and matches it with the total-container-count. 
So in a 3-container job if container-0 fails 3 times, this will mark the job as "succeeded", perhaps not right.